### ISSUE TITLE
More about key reuse

### ISSUE THREAD

**Issue by danvangeest** [lamps-wg/draft-composite-sigs#79]
So, I think I have a use case for key reuse for composite signatures.  To allow the use of the native and alternative private and public keys of Catalyst (hssss) certificates to generate composite signatures.  Composite private and public keys for the certificate wouldn't exist in the wild, but the native and alternative private keys of a certificate could be used to generate a composite signature.

Due to the flexibility of these certificates, this means that a signature could be generated corresponding to the native key, the alternative key, or a composite of the native+alternative key.

Some of you authors already know that I'm not generally in favour of Catalyst certificates, but there are those out there who still want to use them, so I'm coming at this from a harm-reduction perspective.  And also from a forward-looking perspective, this draft is setting the standard for composite signatures.  It would be easier if future composite signatures, for example PQ/PQ, could use the framework of this draft (and implementation) to define other composites (obviously with specific security considerations for the component algorithms involved). At that point, both component algorithms may take a context and key reuse could be done safely.

To that end, I wonder if it would be worth considering enabling key reuse of the component algorithms which take a context parameter by changing the context from:

ctx = Domain

to:

ctx = Domain || Hash(PublicKey1 || PublicKey2)

This would provide domain separation for a key, ML-DSA for example, to be used standalone or as part of multiple composite keys (even multiple composite keys with the same Domain, because the key hash achieves the separation).  I don't claim it's a good idea to reuse a key in multiple composite keys with the same Domain, just that this idea would protect the key if that happened.

I'm interested in your opinions.

**Comment by danvangeest** [comment]
As Scott says elsewhere, 'belt and suspenders` but this is belt and suspenders and superglue.

**Comment by johngray-dev** [comment]
My first thought is what if the signer doesn't have access to the public key?   My second thought is this works great to obtain what you want for ML-DSA or other component algorithms that take a context, but not at all for the ones that don't (which is RSA and most EC combinations except the Edwards ones).  So at best in the context of the current draft it would give us separation for 1 of the components, but not the others, which would foil our security claim of EUF-CMA as long as one of them is EUF-CMA.   My third thought is that we had thought about this property a few years ago (For instance, it would work with multiple independent certs as well), and there probably are lots of interesting use cases where keys could be joined to make a composite signature.  After lots of discussions it seems this property wasn't desired in lieu of the security implications it brought to composite.   

Perhaps for combinations where a context string is used in BOTH algorithms this could make sense...  We had talked about using context strings for other components that could support it.  See  https://github.com/lamps-wg/draft-composite-sigs/issues/64  but it was mostly decided against because the most compatible way is to use no context String...    

So in summary for cases where both components take contexts, I would say it is interesting.    For cases where some components don't take a context String, i don't think we want to go there.

**Comment by danvangeest** [comment]
Here is someone trying to combine separate keys using the composite signature construct.  Check out IPSECME on Monday to see/influence how that goes.
https://datatracker.ietf.org/doc/draft-hu-ipsecme-pqt-hybrid-auth/

Regarding your second thought, yes it protects ML-DSA but not the trad key.  My point is if someone is going to end up doing this anyways, better to protect only the ML-DSA key than neither key.

**Comment by johngray-dev** [comment]
Just looking at this again.   Dan, are you suggesting that if we made this change:

ctx = Domain

to:

ctx = Domain || Hash(PublicKey1 || PublicKey2)

We should allow re-use of the ML-DSA key  (or any other context-supporting algorithm) in some other context?   

An attacker could still construct the above ctx and pass that to a regular ML-DSA verifier (with context) and get the stripped signature validated.  We recently added a prefix to mitigate this issue in traditional verifiers, #86  but I suppose it could be used by a pure ML-DSA verifier to determine if it is coming from a composite.  Then a setting whether a composite validation is being attempted or not would be needed.

**Comment by danvangeest** [comment]
> An attacker could still construct the above ctx and pass that to a regular ML-DSA verifier (with context)

If the attacker can control the context used by a signer or verifier then the context parameter has no value and we shouldn't use it anywhere.  Wherever a signature context is used, it should be a value entirely controlled by the signer or a protocol-specific value.

> My first thought is what if the signer doesn't have access to the public key?

The extra context data doesn't have to be derived from the public key, it just has to be something unique to the composite keypair.  It could even be some random bytes that is encoded into the composite private and public key formats so both signer and verifier have the same value.  The important thing is that it is unique to a single composite keypair.   

>  Dan, are you suggesting that if we made this change: ... We should allow re-use of the ML-DSA key (or any other context-supporting algorithm) in some other context?

I am asking:
- Do you think, despite key reuse being forbidden, people will try to do it?  I think that fact that [draft-hu-ipsecme-pqt-hybrid-auth/](https://datatracker.ietf.org/doc/draft-hu-ipsecme-pqt-hybrid-auth/) already exists suggests yes.
- If we think people will try to do key reuse, should this draft try to protect them as best it can, despite having told them not to do it?
- Alternately, if we have a way to do composites in a way that allows key reuse, should we?

My idea above would not protect ECDSA/RSA from key reuse, but this comment from a different issue is interesting: https://github.com/lamps-wg/draft-composite-sigs/issues/86#issuecomment-2461833968, but was never addressed.  It claims to give strong non-separability, although if the component keys were mixed and matched into different composites I think you could still get some separability-adjacent issues.  If you took that idea and used a unique context per pair of composite keys (public key hashes, or random bytes encoded in the private and public key) then I think you could fix the mix-and-match problem.

**Comment by johngray-dev** [comment]
I'm consolidated the idea in #64 since it was related.  I think it is an interesting idea to add context separation and the lower algorithm level for ALL algorithms.   Today we only add the lower level context separation for ML-DSA...   I reference Illari's idea from https://github.com/lamps-wg/draft-composite-sigs/issues/86#issuecomment-2461833968 as well.

**Comment by johngray-dev** [comment]
I've added the comments for #86 into #64.    I think we should still forbid Key Reuse, but we could implement your idea which was mirror in https://github.com/lamps-wg/draft-composite-sigs/issues/86#issuecomment-2461833968 which I've copied to #64.  Essentially making use of context in all algorithms where it can be used, or wrapping algorithms that don't support context with a SHAKE(M',ctx=composite-sigs).   I think I want to bring this to the mailing list, but I do think it is an interesting idea... 

A couple reasons against it:
- I know some crypto libraries that don't support SHAKE which would now be required to add support for SHAKE BEFORE they could support composite
- Many crypto libraries don't support the context in Ed448 or Ed25519

**Comment by johngray-dev** [comment]
I'm going to bring this to the LAMPS 122 discussion.

**Comment by johngray-dev** [comment]
Hi Dan,

We have discussed this numerous times as an authors group.  We keep coming back to the conclusions
- Timing issue 
  - we don't want to further delay the document by needing further cryptographic review.
- Usability issues: 
  - needing to have the public keys at signing time
  - encoding of the public keys needs to be identical on both sides, which may not always be the case. 

The above are mitigated by the fact that re-use is forbidden.   This seems like a good middle ground compromise.   This is also mitigated by Falco's suggestion of the prefix which if used by a traditional verifier would detect that it should have been a composite signature and not a single signature.

**Comment by johngray-dev** [comment]
In the context of an HSM, the above idea doesn't work as the signer doesn't even have access to the private key... All they have is a reference to the private key.  Therefore any artifacts added to the private key will not be available, meaning the full private key is not available.

I discussed this with Mike Ounsworth today, and probably the best option is to make use of the application specific context if you want to use composites with the Catalyst certificate.   So if we add something like the following to the draft, does this satisfy your request Dan?

"In the event that an application wishes to use two separate keys (for example from two single-algorithm certificates) and use them to construct a single Composite Signature, then it is RECOMMENDED to provide a composite ctx to prevent this signature from being validatable under a composite key made up of the same two component keys.

For example, an application or protocol called Foobar that wishes to do this might invoke the Composite algorithm as:
Composite-ML-DSA.Sign( (sk1, sk2), M ctx="Foobar-dual-cert-sig", PH)"

**Comment by johngray-dev** [comment]
TODO..   I will add this to security considerations....

**Comment by danvangeest** [comment]
I don't think that suggestion would help since the application context being provided is only used to create the Composite-ML-DSA M'.  So a component signature can still be separated from the composite signature to provide a valid single-algorithm signature for M'.  The suggestion would help when using the same composite keypair in two different protocols, but it wouldn't help when using the composite keypair and reusing one of the components.

**Comment by johngray-dev** [comment]
Hi Dan,

The suggestion was for usages outside of this specification like in the catalyst certificate as you mention.  Of course anyone that knows the application specific context could re-construct the message:

M' = prefix | domain | ctx(len) | ctx | HashOID | PH (r | Message)  

As of this week, they would also need the r from the full composite signature to make the construction before they could verify.  At that point if they are doing all this message work to verify one component of the composite signature, they are basically purposefully doing a partial composite validation but calling it a single validation.

**Comment by johngray-dev** [comment]
Would putting the application context or possibly r into the ML_DSA context help?    Lets think about it.

**Comment by johngray-dev** [comment]
Thanks Dan,

I agree, adding the r into the context gives the attacker some control over the context string, which could be manipulated.

So we are left with this:

> Reusing the ML-DSA key in another composite key

The r does help with some subset of key reuse, but not all cases.   It solves we call the "mixed-key forgery attack": Take two composite keys `(mldsaPK1, tradPK1)` and `(mldsaPK2, tradPK2)` which do not share any key material and have them produce signatures `(r1, mldsaSig1, tradSig1)` and `(r2, mldsaSig2, tradSig2)` respectively over the same message `M`. Consider whether it is possible to construct a forgery by swapping components and presenting `(r, mldsaSig1, tradSig2)` that verifies under a forged public key `(mldsaPK1, tradPK2)`. This forgery attack is blocked by the randomizer `r` so long as `r1 != r2`.

Also if we have this more conventional key re-use scenario

Keya = MldsaA+TradA1
Keyb = MldsaA+TradA2

Any signature produced by Key1 or Key2 would be of the form
(r1, siga1,siga2) -> siga
(r2, sigb1, sigb2)  ->sigb

M1 = prefix | domain | ctx(len) | ctx | HashOID | r1 |  PH (r1 | Message)
M2 = prefix | domain | ctx(len) | ctx | HashOID | r2 |  PH (r2 | Message)

If you try to verify siga using the sigb key (as a composite), it will fail thanks to the traditional component, but that doesn't prevent key re-use:
You need to compute:
M1 = prefix | domain | ctx(len) | ctx | HashOID | r1 |  PH (r1 | Message)
Keyb.mldsa would be able to replace Keyb.mldsa (its the same key and ctx domain), but the traditional component would still cause an overall failure.  

- if r1 was passed down into the ML-DSA ctx, so that MLDSA.ctx = domain | r1 then:

MLDSA.Verify(Keyb.MldsaA, M1,ctx=domain | r1) = MLDSA.Verify(Keya.MldsaA, M1, ctx=domain | r1)

So it wouldn't help at all.

- if Hash (Keya) is passed down into the ML-DSA ctx, so that MLDSA.ctx = domain | Hash (Keya) - the original request:

k = Hash(Keya)
MLDSA.Verify(Keyb.MldsaA, M1, ctx=domain | k) != MLDSA.Verify(Keya.MldsaA, M1, ctx=domain | Hash(Keyb) )

This works because you can do a check on the full public key being used to verify the signature at that instance.  I think the only solution is probably using the Hash of the public key (or some combined aspects of the key itself that can be checked by the honest composite verifier).   Thus we have come full circle.  

So unfortunately, given the design constraint of not wanting to require access to the public key at signing time, and the fact that the private key components may not always be available in bytes but only in a reference on a HSM, I think we are left with the very strong language against key re-use in the draft.
